### PR TITLE
AeroDyn driver update for multiple wind turbines, with arbitrary motions and geometries

### PR DESCRIPTION
**Pull request by ebranlard**
This pull request is almost ready to merge. 

**Feature or improvement description**
New AeroDyn driver functionalities:
- Multiple turbines with arbitrary number of blades and orientation (advanced rotor inputs)
- Different motions for the base, nacelle yaw, and individual blade pitch: fixed values, sinusoidal, or arbitrary via time series 
- Simple wind definition (Power law), or InflowWind 

**Old features still supported:**
- Basic rotor inputs for HAWT (simply provide HubHeight, Tilt, Tower2Shaft)
- Prescribed Time series of Wind Speed / PL Exp / Pitch / Yaw  (`AnalysisType=2`)
- Combined case analyses (`AnalysisType=3`)

**Example of Applications**
- Sinusoidal motions of the Platform base
- Vertical axis turbines
- Multiple wind turbines
- Quad rotor
- Kites

110567794-4c540780-810f-11eb-9cf2-23273877ec0f

**Impacted areas of the software**
AeroDyn, OLAF

**Additional supporting information**

- [Live documentation](https://ebranlard-openfast.readthedocs.io/en/f-driver/source/user/aerodyn/driver.html)  (will be updated soon)
-  [Test/examples files](https://github.com/OpenFAST/r-test/tree/f/driver/modules/aerodyn) files for all the different features with README describing each cases.

**Check list**

- [x] Merge #672 
- [x] Merge #589
- [x] Merge #648
- [x] Finalize documentation
- [x] Merge/update new r-tests
- [x] Review
Extra features:
- [x] WrVTK harmony with OpenFAST
- [x] InitRot before debug outputs

**Test results, if applicable**
Expected some small differences for OLAF Elliptical wing, will be resolved after #648 is merged

### CODE REVIEW

File: docs/source/user/aerodyn/driver.rst [hunk 1]
```diff
@@ -0,0 +1,497 @@
...
+
+The user specify the number of turbines as follows:
+
+.. code:: 
+
+    ----- Turbine Data --------------------------------------------------------------
+    1   NumTurbines  - Number of turbines (should be 1 for AnalysisType=2 or AnalysisType=3)
+
+As noted in the comment, the number of turbine should be 1 for `AnalysisType=2` and `AnalysisType=3`.
+After this section, the geometry and motion is provided for each turbine. Inputs for each turbine must have the suffix `(i)` where `i` is the turbine number.
```
Comment (jjonkman) [inline docs/source/user/aerodyn/driver.rst:119]: Add "(even with NumTurbines = 1, then i=1)"

File: docs/source/user/aerodyn/driver.rst [hunk 2]
```diff
@@ -0,0 +1,497 @@
...
+Outputs for each turbine will be written to distinct files, with the suffix `.Ti` where `i` is the turbine number.
+
+An example of configuration with two wind turbines is shown in :numref:`fig:MultiRotor`. The figure defines the different frames and origin associated with each turbine: the turbine base frame (t), nacelle frame (n), hub frame (h), and blade frames (b). The notations and conventions follow the OpenFAST frames, except that the turbine frame does not have its origin at the tower base.
+Prescribed motions of the turbine occur at the turbine origin.
+Yawing occurs around the :math:`z_n` axis,  the blade rotates about the :math:`x_h` axis, and pitching occurs around the individual :math:`z_b` axes. The definitions of the different frames are standardized when using a basic (HAWT) input format definition, and are arbitrarily defined when using the advanced input format. More details are given in the next paragraph.
```
Comment (jjonkman) [inline docs/source/user/aerodyn/driver.rst:124]: Change "blade rotates" to "rotor rotates" and add "blade" before pitching.

File: docs/source/user/aerodyn/driver.rst [hunk 3]
```diff
@@ -0,0 +1,497 @@
...
+   Definition of multiple rotors. 
+
+
+
+
+**Turbine geometry definition**
+
+Two turbine input formats are supported:
+
+- basic (`BasicHAWTFormat=True`): Basic horizontal axis wind turbine (HAWT) format. In this format, the turbine geometry is entirely determined by the number of blades (`NumBlades`), the hub radius (`HubRad`), the hub height (`HubHt`), the overhang (`Overhang`), the shaft tilt (`ShftTilt`) and the precone (`Precone`). The definition of each parameter follows the OpenFAST convention. Additionally, the user needs to provide the origin of the turbine base at `t=0` (`BaseOriginInit`).
```
Comment (jjonkman) [inline docs/source/user/aerodyn/driver.rst:139]: I would say OpenFAST does not have a convention; perhaps you mean "ElastoDyn convention"?

File: docs/source/user/aerodyn/driver.rst [hunk 4]
```diff
@@ -0,0 +1,497 @@
...
+              3.    HubRad(1)       - Hub radius (m)
+          140.82513 HubHt(1)        - Hub height (m)
+              -7    Overhang(1)     - Overhang (m)
+              -6    ShftTilt(1)     - Shaft tilt (deg)
+              -4    Precone(1)      - Blade precone (deg)
+
+
+- advanced (`BasicHAWTFormat=False`): The position and orientation of the tower base, nacelle, hub, and individual blades can be arbitrarily defined. This can be used for HAWT and any other turbine concepts. 
+  The definition of the different frames are given in :numref:`fig:MultiRotor`.
+  The position (`BaseOriginInit`) and orientation (`BaseOriginInit`) of the turbine base frame are defined with respect to the global frame. Orientations are given using the values of three successive rotations (x-y-z Euler angles). If the base undergoes a motion, the orientation of the base frame will consists of the time-varying rotations followed by these initial rotations.
```
Comment (jjonkman) [inline docs/source/user/aerodyn/driver.rst:157]: Change "angles" to "angle sequence".
Change "consists" to "consist".

File: docs/source/user/aerodyn/driver.rst [hunk 5]
```diff
@@ -0,0 +1,497 @@
...
+
+  A flag indicating whether the turbine has a tower is given on the next line (`HasTower`). This flag currently affects the VTK outputs and does not have an impact on AeroDyn yet. The user still has to provide tower inputs data in AeroDyn for each turbine (see :numref:`ad_inputs_multirot`).
+  The next line indicates which projection AeroDyn is to use in its calculation. It is recommended to use `HAWTProjection=True` for HAWT, which is the default projection used in OpenFAST (projects on the coned-pitched axis). For other rotor concepts, set `HAWTprojection=False`.
```
Comment (jjonkman) [inline docs/source/user/aerodyn/driver.rst:160]: Add "BEMT" before "calculation".
Change "OpenFAST" to "AeroDyn".

File: docs/source/user/aerodyn/driver.rst [hunk 6]
```diff
@@ -0,0 +1,497 @@
...
+  The following lines indicates the position and orientations of the tower, nacelle and hub.
```
Comment (jjonkman) [inline docs/source/user/aerodyn/driver.rst:161]: Change "indicates" to indicate".

File: docs/source/user/aerodyn/driver.rst [hunk 7]
```diff
@@ -0,0 +1,497 @@
...
+  The next line indicates which projection AeroDyn is to use in its calculation. It is recommended to use `HAWTProjection=True` for HAWT, which is the default projection used in OpenFAST (projects on the coned-pitched axis). For other rotor concepts, set `HAWTprojection=False`.
+  The following lines indicates the position and orientations of the tower, nacelle and hub. 
+
+  The tower and the nacelle are defined with respect to the turbine base (t) origin and frame.
+  The tower top is assumed to coincide with the nacelle origin. 
+  The tower stations defined in the AeroDyn input file are assumed to be given with respect to the tower origin, unlike OpenFAST which uses ground/MSL as a reference (see :numref:`ad_inputs_multirot`).
+  The hub is defined with respect to the nacelle origin and frame (n).
+
+  The definitions of the blades follows, starting with the number of blades `NumBlades`. A rotor with zero blade is supported and can be used to model an isolated tower.
```
Comment (jjonkman) [inline docs/source/user/aerodyn/driver.rst:168]: Change "follows" to follow".
Change "blade" to "blades".

File: docs/source/user/aerodyn/driver.rst [hunk 8]
```diff
@@ -0,0 +1,497 @@
...
+  If tower shadow/potential is used in AeroDyn, then the isolated tower will disturbed the flow of the vortex wake when OLAF is used.
```
Comment (jjonkman) [inline docs/source/user/aerodyn/driver.rst:169]: Change "disturbed" to "disturb".

File: docs/source/user/aerodyn/driver.rst [hunk 9]
```diff
@@ -0,0 +1,497 @@
...
+  When BEM is used, the flow of the blades of a given turbine are disturbed only by the turbine's tower.
```
Comment (jjonkman) [inline docs/source/user/aerodyn/driver.rst:170]: Change "the turbine's" to "that turbine's"

File: docs/source/user/aerodyn/driver.rst [hunk 10]
```diff
@@ -0,0 +1,497 @@
...
+  The inputs for turbine `i` and blade `j` are labelled `(i_j)`.
+  The origin `BldOrigin_h`) and orientation (`BldOrientation_h`) of each blade are given with respect to the hub origin and frame (h).
```
Comment (jjonkman) [inline docs/source/user/aerodyn/driver.rst:172]: Add "(" before BldOrigin_h.

File: docs/source/user/aerodyn/driver.rst [hunk 11]
```diff
@@ -0,0 +1,497 @@
...
+
+The base motion is given in the same way for basic or advanced geometries.
+The motion of the base may be: fixed (`BaseMotionType=0`), sinusoidal (`BaseMotionType=1`) or arbitrary (`BaseMotionType=2`). 
+The turbine base motion is applied at each time step before applying the initial position and orientation of the turbine base.
+A sinusoidal motion implies that one degree of freedom (`DegreeOfFreedom`) of the turbine base is moving according to a sine function of a given amplitude (`Amplitude`) and frequency (`Frequency`), with zero phase.
+The 6 possible degrees of freedom correspond to translations or rotations of the base frame in global coordinates (g) (e.g. surge, sway, heave, roll, pitch, yaw).
+An arbitrary motion is specified via a CSV file (`BaseMotionFileName`) which contains 19 columns: time, 3 translations (global), three successive rotations (global), 3 translation velocities, 3 rotational velocities (omega, in global), 3 translational accelerations and 3 rotational accelerations (alpha, in global). Example of arbitrary input files are given in :numref:`ad_inputfiles_examples`.
```
Comment (jjonkman) [inline docs/source/user/aerodyn/driver.rst:212]: Add "The displacements/orientations, velocities, and accelerations are not checked internally for consistency."

File: docs/source/user/aerodyn/driver.rst [hunk 12]
```diff
@@ -0,0 +1,497 @@
...
+
+
+**Turbine motion definition**
+
+The definition of the turbine motion are only used when `AnalysisType=1`, but must always be present in the input file. 
+
+The base motion is given in the same way for basic or advanced geometries.
+The motion of the base may be: fixed (`BaseMotionType=0`), sinusoidal (`BaseMotionType=1`) or arbitrary (`BaseMotionType=2`). 
+The turbine base motion is applied at each time step before applying the initial position and orientation of the turbine base.
+A sinusoidal motion implies that one degree of freedom (`DegreeOfFreedom`) of the turbine base is moving according to a sine function of a given amplitude (`Amplitude`) and frequency (`Frequency`), with zero phase.
```
Comment (jjonkman) [inline docs/source/user/aerodyn/driver.rst:210]: Is "Frequency" in rad/s or Hz?

File: docs/source/user/aerodyn/driver.rst [hunk 13]
```diff
@@ -0,0 +1,497 @@
...
+  Examples are given below:
+
+.. code::
+
+    0         NacYaw(1)        - Yaw angle (about z_t) of the nacelle (deg)
+    7         RotSpeed(1)      - Rotational speed of rotor in rotor coordinates (rpm)
+    1         BldPitch(1)      - Blades pitch (deg)
+
+- advanced: When an advanced geometry is provided and when the number of blades is non-zero, the motion section contains options for the nacelle motion, rotor motion and individual blade pitch motion. 
+  The syntax for each of these motion consists of defining a type (fixed or time-varying), a value for the fixed case or a file for the time-varying case.
```
Comment (jjonkman) [inline docs/source/user/aerodyn/driver.rst:239]: Change "motion" to "motions".

File: docs/source/user/aerodyn/driver.rst [hunk 14]
```diff
@@ -0,0 +1,497 @@
...
+  The input files are CSV files containing time, position, speed and acceleration. Examples of files are given in :numref:`ad_inputfiles_examples`.
+  If the pitch rate, yaw rate, and accelerations are not known, the user should input zeros in these columns.
```
Comment (jjonkman) [inline docs/source/user/aerodyn/driver.rst:241]: I'm not sure this needs to be stated (may cause confusion).  Instead, I would say--as above--"The displacements/orientations, velocities, and accelerations are not checked internally for consistency."

File: docs/source/user/aerodyn/driver.rst [hunk 15]
```diff
@@ -0,0 +1,497 @@
...
+  The time vector in the motion file has to be ascending, but does not need to be linear. Linear interpolation is used by the driver to determine inputs at a given time.
```
Comment (jjonkman) [inline docs/source/user/aerodyn/driver.rst:242]: These comments should added to the base motion section as well.

File: docs/source/user/aerodyn/driver.rst [hunk 16]
```diff
@@ -0,0 +1,497 @@
...
+    "unused"  BldMotionFileName(1_1) - Filename containing blade pitch motion [used only when BldMotionType=1]
+    "unused"  BldMotionFileName(1_2) - Filename containing blade pitch motion [used only when BldMotionType=1]
+    "unused"  BldMotionFileName(1_3) - Filename containing blade pitch motion [used only when BldMotionType=1]
+
+  
+
+
+**Time-dependent analysis**
+
+Time-dependent analyses are used to vary few standard variables during the simulation.
```
Comment (jjonkman) [inline docs/source/user/aerodyn/driver.rst:267]: Add "a" before "few".

File: docs/source/user/aerodyn/driver.rst [hunk 17]
```diff
@@ -0,0 +1,497 @@
...
+The output section controls the format of the tabular output file and VTK files, similar to the OpenFAST outputs.
+The user can control the hub radius and nacelle dimension for the VTK visualization. The nacelle is represented as a sphere of radius (`VTKHubRad`), and the nacelle with a parallelepiped defined using an origin and three lengths parallel to the nacelle coordinates (`VTKNacDim`).
+
+
+.. code::
+
+    ----- Output Settings -------------------------------------------------------------------
+      "ES15.8E2"     OutFmt      - Format used for text tabular output, excluding the time channel.  Resulting field should be 10 characters. (quoted string)
+    2                OutFileFmt  - Format for tabular (time-marching) output file (switch) {1: text file [<RootName>.out], 2: binary file [<RootName>.outb], 3: both}
+    0                WrVTK       - VTK visualization data output: (switch) {0=none; 1=animation}
```
Comment (jjonkman) [inline docs/source/user/aerodyn/driver.rst:311]: Perhaps this doesn't matter, but OpenFAST uses WrVTK = 1 for initialization data only and 2 for animation

File: docs/source/user/aerodyn/driver.rst [hunk 18]
```diff
@@ -0,0 +1,497 @@
...
+To minimize the impact of the multiple-turbines implementation, the driver currently uses only one AeroDyn input file for all turbines. 
+This means that the AeroDyn options are currently the same for all rotors.
+
+The definition of the blade files and tower inputs needs to be adapted when more than three blades are used and more than one turbine is used.
+
+**Blade files**
+
+The legacy AeroDyn format requires a minimum of three blade file names. 
+For this reason, the blades of all rotors are currently indicated successively in the `ADBlFile` list. 
+The list is populated by looping on turbines and turbine blades, with tine blades being the fast index.
```
Comment (jjonkman) [inline docs/source/user/aerodyn/driver.rst:334]: What is "tine"?  Suggest rewording.

**Review by jjonkman — COMMENTED** [review]
Thanks @ebranlard.  Overall, this looks like a powerful extension to the AeroDyn driver.  I've spent some time reviewing the updated documentation.  Please find my edits and comments here.

File: docs/source/user/aerodyn/driver.rst [hunk 19]
```diff
@@ -0,0 +1,497 @@
...
+            0.5     DT           - Simulation time step [used only when AnalysisType/=3] (s)
+    "AD.dat"        AeroFile     - Name of the primary AeroDyn input file
+
+
+
+**Inflow data**
+
+The inflow can be provided in two ways:
+
+- basic (`CompInflow=0`): uniform wind with a power law shear. The wind is defined using a reference height (`RefHt`), a powerlaw exponent (`PLExp`), and the wind speed at reference height (`HWindSpeed`). With `AnalysisType=1`, the reference wind speed and power law are defined separately as time series (see "time-dependent analysis"). With `AnalysisType=3`, these parameters are provided in a separate table (see "Combined-Case analyses"). The reference height is used for all analyses types, since this height may be different than the hub height.
```
[the remark this answers is not in the document]
Comment (ebranlard) [reply docs/source/user/aerodyn/driver.rst:93]: Yes, it was a typo, with AnalysisType=2 or 3, PLExp and HWindSpeed are not used.

File: docs/source/user/aerodyn/driver.rst (hunk 5 above)
Comment (ebranlard) [reply docs/source/user/aerodyn/driver.rst:160]: Some of the outputs of AeroDyn rely on this projection convention (BEMT or FVW).

File: docs/source/user/aerodyn/driver.rst (hunk 18 above)
Comment (ebranlard) [reply docs/source/user/aerodyn/driver.rst:334]: I'm not sure what happened there.. The blade index is the fastest index.

File: docs/source/user/aerodyn/driver.rst (hunk 17 above)
Comment (ebranlard) [reply docs/source/user/aerodyn/driver.rst]: changed to match openfast convention now

File: modules/aerodyn/src/AeroDyn.f90 [hunk 20]
```diff
@@ -2780,145 +2806,76 @@ SUBROUTINE Init_FVWmodule( InputFileData, u_AD, u, p, x, xd, z, OtherState, m, E
...
-   character(*), parameter                       :: RoutineName = 'Init_FVWmodule'
-
-   ! note here that each blade is required to have the same number of nodes
-
+   integer(intKi)                               :: nWings         ! total number of wings
+   integer(intKi)                               :: j              ! node index
+   integer(intKi)                               :: iR, iB, iW, iW_incr  ! rotor index, blade index, wing index
+   integer(IntKi)                               :: ErrStat2
+   character(ErrMsgLen)                         :: ErrMsg2
+   character(*), parameter                      :: RoutineName = 'Init_FVWmodule'
```
Comment (andrew-platt) [inline modules/aerodyn/src/AeroDyn.f90:2814]: Change routinename to `Init_OLAF`

File: modules/aerodyn/src/AeroDyn_IO.f90 [hunk 21]
```diff
@@ -1786,19 +1878,19 @@ subroutine Calc_WriteOutput_BEMT
...
-!         m%AllOuts( BAeroFx(k) ) = tmp(1)
-!         m%AllOuts( BAeroFy(k) ) = tmp(2)
-!         m%AllOuts( BAeroFz(k) ) = tmp(3)
-!         ! Transform moment vector to blade root coordinate system
-!         tmp = matmul( u%BladeRootMotion(k)%Orientation(:,:,1), m%BladeRootLoad(k)%moment( :,1) )
-!         m%AllOuts( BAeroMx(k) ) = tmp(1)
-!         m%AllOuts( BAeroMy(k) ) = tmp(2)
-!         m%AllOuts( BAeroMz(k) ) = tmp(3)
-!      end do  ! k=blades
+      do k=1,min(p%NumBlades,4)
```
Comment (rafmudaf) [inline modules/aerodyn/src/AeroDyn_IO.f90:1881]: @ebranlard Why the loop to a minimum of 4 here? In other areas, the minimum is 3. Also, using `MaxBl` allows to change it all at once and prevent any inconsistencies.

File: modules/aerodyn/src/AeroDyn_IO.f90 [hunk 22]
```diff
@@ -1926,6 +2042,21 @@ subroutine Calc_WriteOutput_FVW
         m%AllOuts( RtAeroCt ) = m%AllOuts( RtAeroFxh ) /  denom
       end if
 
+      ! Integrate force/moments over blades by performing mesh transfer to blade root points:
+      do k=1,min(p%NumBlades,4)
```
Comment (rafmudaf) [inline modules/aerodyn/src/AeroDyn_IO.f90:2046]: Same here, see above

File: modules/aerodyn/src/AeroDyn_IO.f90 (hunk 21 above)
Comment (rafmudaf) [reply modules/aerodyn/src/AeroDyn_IO.f90:1881]: @andrew-platt added some clarity in that these outputs are specific to VAWTS. Without context, the difference in 3 and 4 is unclear, so a comment noting this would be helpful.

Comment (ebranlard) [reply modules/aerodyn/src/AeroDyn_IO.f90:1881]: The limit is purely arbitrary for now, and not limited to VAWT. I needed more outputs (actually I needed 9 blades), so I hardcoded outputs for an extra blade (without using the matlab script). Hannah's branch will have a better handling of the outputs, this is just temporary.

**Comment by rafmudaf** [comment]
@ebranlard you resolve the merge conflict with the inflow wind file, I'll deal with r-test.

**Comment by andrew-platt** [comment]
I looked at the conflict and couldn't figure out why git is complaining about it.

**Comment by ebranlard** [comment]
It seems it was just a couple of spaces on one line. I'm not sure if this branch is pointing to the right r-tests after this merge though..

**Comment by rafmudaf** [comment]
I think we still need to merge the r-test branch used here into that repository's `dev` branch so that we don't undo the update to r-test in https://github.com/OpenFAST/openfast/commit/792f37ee660d507191ae0807038bb791a4430c3b. Is that right?

**Comment by ebranlard** [comment]
I've updated the r-tests, merged with the latest r-test-dev and updated the vertical axis testcase. 
I believe this pull request is ready to be merged. Maybe the r-test can be squashed?

**Review by andrew-platt — CHANGES_REQUESTED** [review]
I think it all looks good, except the detail in `IfW_FFWind_Base.f90`

File: modules/inflowwind/src/IfW_FFWind_Base.f90 [hunk 23]
```diff
@@ -485,6 +485,10 @@ SUBROUTINE GetInterpValues()
                RETURN
             ENDIF
          ENDIF
+         if (IYLO<0) then
+            STOP
```
[the remark this answers is not in the document]
Comment (ebranlard) [reply modules/inflowwind/src/IfW_FFWind_Base.f90]: I probably introduced this while debugging the error handling that was missing above. I can remove it now.

**Review by ebranlard — COMMENTED** [review]
I've merged with the latest dev and applied the changes requested.

**Review by andrew-platt — APPROVED** [review]

**Comment by andrew-platt** [comment]
@ebranlard suggested this be merged after #729.
